Add OIDC support #494

louise-davies opened the pull request:
## Description

Adds support for OIDC via the icat authenticator, and adds 2 new authenticators: `oidc` and `password_and_oidc` - the former is for purely OIDC and the latter allows for both user/pass and OIDC( e.g. OG)

Also, majorily refactored LoginPage and the auth providers so that the fetching of e.g. ICAT mnemonics is now in the authenticators not LoginPage. Plus extracted out a lot of common code to `baseAPIAuthProvider` e.g. verify and refresh, OIDC login stuff, maintenance endpoints etc.

Only other big notable for testing is that I added a docker compose to set up a suitable keycloak e.g. for e2e tests, but since I added the code to `auth-server.js` this can also be used to test SG in dev mode with e.g. either the `oidc` or `userpass_and_oidc` authenticators

Also, there are breaking changes due to how OG-api and sg-auth handled the maintenance endpoints
- They now expect PUT requests rather than POST, which affects OG-api
- They now expect the token to be in the header and the request body to be the new state, affecting sg-auth

## Testing instructions
Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] Verify OIDC workflow works with new authenticators
- [ ] Verify OIDC workflow works with ICAT authenticator (will need to speak to me about setting up a suitable sg-auth server)

## Agile board tracking
Closes #494

louise-davies committed:
Add initial ICAT OIDC working code

louise-davies committed:
#494 - refactor multiple authenticator code to be primarily located in the auth providers not the login page

louise-davies committed:
Add passwordAndOIDCAuthProvider (for OG) and extract a lot of common logic to new baseAPIAuthProvider

louise-davies committed:
#494 - fix some bugs with the OIDC login
- ensure we set up OIDC on login fail to set up for a new attempt
- ensure that initialising a single authenticator doesn't overwrite OIDC vars
- no longer append oidc_ to oidc logins and instead just use the configuration url as key
 - icat auth has other ways to identify an OIDC authenticator

louise-davies committed:
#494 - add oidc endpoints to auth-server.js
- requires a keycloak/other OIDC provider at localhost:8081

louise-davies committed:
Improve login page styling

louise-davies committed:
Add unit tests for OIDC login code

louise-davies committed:
#494 - add e2e tests for OIDC code

louise-davies committed:
Merge branch 'develop' into feature/icat-oidc-auth-#494

joelvdavies commented on cypress/e2e/login.cy.ts:
```diff
@@ -439,6 +439,34 @@ describe('Login', () => {
       cy.get('[aria-label="Open user menu"]').should('be.visible');
     });
 
+    it('should be able to login via SSO after auto login and be displayed as logged in', () => {
+      cy.visit('/login');
+
+      cy.get('#select-mnemonic').click();
+      cy.contains('Keycloak').click();
+
+      cy.url().as('originUrl', { type: 'static' });
+
+      cy.contains('button', 'Login with Keycloak').click();
+
+      cy.url().should('include', 'http://localhost:8081');
+      // login to keycloak
+      cy.origin('http://localhost:8081', () => {
+        cy.get('#username').type('test');
+        cy.get('#password').type('password');
+        cy.get('#kc-login').click();
+      });
+
+      // redirect?
+      // cy.origin()
+      cy.get('@originUrl').then((url) => {
+        cy.url().should('include', url);
+      });
+
+      cy.contains('Sign out').should('exist');
+      cy.get('[aria-label="Open user menu"]').should('be.visible');
+    });
```
This is repeated in the describe below.

louise-davies replied:
They are meant to be different test cases, one is testing the plain OIDC auth and one is testing it working nicely with the `autoLogin` feature - however I forgot to rename the second test!

joelvdavies commented on package.json:
```diff
@@ -32,6 +32,7 @@
     "i18next-http-backend": "2.7.3",
     "js-cookie": "3.0.1",
     "jsonwebtoken": "9.0.0",
+    "jwks-rsa": "3.2.0",
```
Should this be a dev dependency as its only used in the auth-server?

joelvdavies commented on server/auth-server.js:
```diff
@@ -83,6 +84,80 @@ app.post(`/login`, function (req, res) {
+  res.status(200).json(accessToken);
+});
+
+// Fetch Maintenance State
```
I don't think this comment makes sense?

joelvdavies commented on src/authentication/baseAPIAuthProvider.tsx:
```diff
@@ -0,0 +1,280 @@
+            type: NotificationType,
+            payload: {
+              message:
+                'It is not possible to authenticate you at the moment. Please, try again later',
```
```suggestion
                'It is not possible to authenticate you at the moment. Please, try again later.',
```

joelvdavies commented on src/authentication/baseAPIAuthProvider.tsx:
```diff
@@ -0,0 +1,280 @@
+    });
+}
+
+// GENERATING CODE VERIFIER
```
Just wondering where this comes from and its context in OIDC?

joelvdavies commented on src/state/state.types.tsx:
Unrelated to this PR but is the Plugin interface up to date? I didn't think unauthorised is in the plugin config but the routes?

louise-davies replied:
What do you mean by this? both the `RegisterRoutePayload` and the `PluginConfig` have `unauthorised` in them?

joelvdavies replied:
I mean here, https://github.com/ral-facilities/scigateway/blob/b92311be63b4835458742e39cad07cd78ff7543d/src/state/state.types.tsx#L13-L19. I didn't think it was in the plugin config in the settings.json just the routes in the individual plugin configs.

louise-davies replied:
Ah, I think this may have been a holdover from when we had the idea of allowing a whole plugin to be unauthorised like this, but yes this option doesn't do anything so I'll remove it

joelvdavies reviewed (commented):
I think the description needs updating as it references `password_and_oidc` which is now `userpass_and_oidc`.

The existing authenticators worked well for me for local instances of IMS, DG and OG. I also tried username and password using `userpass_and_oidc` for OG which worked as well. The tests all worked locally for me too.  Have not tried OIDC directly yet.

joelvdavies reviewed (commented):
Just tried the oidc providers with the local keycloak instance and auth server which worked well 🥳 . I assume it would be tricky in practice to get a sensible username to display here?
<img width="373" height="141" alt="image" src="https://github.com/user-attachments/assets/842f03c4-58a6-4727-9121-06343f8fcdda" />

louise-davies committed:
Move packages from dependencies to devDependencies and remove unneeded packages

louise-davies committed:
Address PR comments

louise-davies commented:
> Just tried the oidc providers with the local keycloak instance and auth server which worked well 🥳 . I assume it would be tricky in practice to get a sensible username to display here? <img alt="image" width="373" height="141" src="https://private-user-images.githubusercontent.com/90245114/469765040-842f03c4-58a6-4727-9121-06343f8fcdda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzNjEzMTAsIm5iZiI6MTc1MzM2MTAxMCwicGF0aCI6Ii85MDI0NTExNC80Njk3NjUwNDAtODQyZjAzYzQtNThhNi00NzI3LTkxMjEtMDYzNDNmOGZjZGRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI0VDEyNDMzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5ZjA4OWMzODkxNTFhYWU0NTNkNDU0NWI0Mzc2ZTdlOTVkOTRkY2MwYzdiYjA2MmNiYjA3NDcyZmM0Y2FjZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OfboHn6Gkd9gktza0svAppcGtTgwlK5nbQ5Efx4Gn48">

it's up to the auth server to put something sensible in the username field - I didn't put too much thought into the `auth-server.js` implementation tbh but in proper implementations we would expect something more like the user's email or fed id yeah.

I've changed it to attempt to read the email from the token if it's there :)

louise-davies committed:
Remove unused option from Plugin interface
- it never did anything

louise-davies committed:
Merge branch 'develop' into feature/icat-oidc-auth-#494

louise-davies committed:
Add full stop to error message to match other error messages

joelvdavies reviewed (dismissed):
LGTM 🥳

joshdimanteto reviewed (dismissed):
LGTM

louise-davies committed:
Merge branch 'develop' into feature/icat-oidc-auth-#494

louise-davies committed:
Add default messages to display when scheduled maintenance/maintenance state is updated
- some APIs may not opt to return a success message (e.g. OG) so have a fallback

louise-davies committed:
Add email and profile scopes to OIDC authorisation request

louise-davies committed:
Fix logout & re-login OIDC bug & fix if condition for non-multiple authenticators
- need to ensure we properly reset any OIDC variables on mount if we're not coming back from a redirect
- the latter was causing render issues when a multiple authenticator first rendered with no authenticator set

louise-davies committed:
Merge branch 'develop' into feature/icat-oidc-auth-#494

louise-davies committed:
Add support for non-PKCE workflows
- plus change some extra stuff to account for Alan's changes

louise-davies committed:
Improve OIDC security by using state and nonce variables & fix referrer redirect when using OIDC

louise-davies committed:
Fix/add unit tests due to OIDC changes

louise-davies committed:
Update e2e tests to test OIDC non-pkce workflow

louise-davies commented:
Added support for the updated interface from Alan, tested against sg-auth oidc branch and v0.7 of OG (this branch is deployed and working on OG preprod)

louise-davies committed:
Remove setTimeout workaround for invalid "future" iat issue

joelvdavies commented on src/authentication/baseAPIAuthProvider.tsx:
```diff
@@ -0,0 +1,404 @@
+      // need to talk to backend people about it
+      await new Promise<void>((resolve, _reject) => {
+        setTimeout(() => resolve(), 1_000);
+      });
```
I think this might have been addressed?
I can see Alan added a leeway on ral-facilities/scigateway-auth@aca9fab
Alex has on ral-facilities/operationsgateway-api@1ee23e6
and Viktor on https://github.com/ral-facilities/ldap-jwt-auth/pull/272/commits/073275370b2e59f71c24bae0a16f0de8a8a448fe.

louise-davies replied:
Oh I must have forgot to push the commit 😂

joelvdavies reviewed (commented):
Just a query, worked well with IMS and https://github.com/ral-facilities/ldap-jwt-auth/pull/272.
